### PR TITLE
WIP: Topics/testing

### PR DESCRIPTION
**Pull request by zakalawe**
Current state of the test-harness and driver, for feedback and discusson. @Marcus10110

### CODE REVIEW

**Comment by zakalawe** [comment]
@Marcus10110 pushed some more commits, tests are running now so it's a good time to get feedback and as in-depth a code reivew as you can do.

File: AnalyzerTestHarness/SettingsStubs.cpp [hunk 1]
```diff
@@ -0,0 +1,247 @@
+#include "AnalyzerSettings.h"
+#include "AnalyzerSettingInterface.h"
+
+#include <cmath>
```
Comment (Marcus10110) [inline AnalyzerTestHarness/SettingsStubs.cpp]:
```
#include <cassert>
#include <algorithm>
```

File: AnalyzerTestHarness/MockSimulatedChannelDescriptor.cpp [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+}
+
+void SimulationChannelDescriptor::Advance(U32 num_samples_to_advance)
+{
+
+}
+
+BitState SimulationChannelDescriptor::GetCurrentBitState()
+{
+
```
Comment (Marcus10110) [inline AnalyzerTestHarness/MockSimulatedChannelDescriptor.cpp]: MSVC requires this to return a value.

File: AnalyzerTestHarness/MockSimulatedChannelDescriptor.cpp [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+}
+
+U64 SimulationChannelDescriptor::GetCurrentSampleNumber()
+{
```
Comment (Marcus10110) [inline AnalyzerTestHarness/MockSimulatedChannelDescriptor.cpp]: MSVC requires this to return a value.

File: AnalyzerTestHarness/HelperStubs.cpp [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+    std::string result = ss.str();
+    if( ( result.size() + 1 ) > result_string_max_length )
+        result = result.substr( 0, result_string_max_length - 1 );
+
+    strcpy( result_string, result.c_str() );
+}
+
+U64 AnalyzerHelpers::AdjustSimulationTargetSample(U64 target_sample, U32 sample_rate, U32 simulation_sample_rate)
+{
+
```
Comment (Marcus10110) [inline AnalyzerTestHarness/HelperStubs.cpp]: MSVC requires this to return a value.

File: AnalyzerTestHarness/HelperStubs.cpp [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+    delete mData;
+    mData = NULL;
+}
+
+void SimpleArchive::SetString( const char* archive_string )
+{
+ }
+
+const char* SimpleArchive::GetString()
+{
```
Comment (Marcus10110) [inline AnalyzerTestHarness/HelperStubs.cpp]: MSVC requires this to return a value.

File: AnalyzerTestHarness/TestInstance.cpp [hunk 6]
```diff
@@ -0,0 +1,141 @@
+#include "TestInstance.h"
+
+#include "AnalyzerResults.h"
+#include "TestAnalyzerData.h"
+#include "MockSimulatedChannelDescriptor.h"
+
+#include <cassert>
+#include <iostream>
```
Comment (Marcus10110) [inline AnalyzerTestHarness/TestInstance.cpp]:
```
#include <cstring> //for memcpy on gcc
```

File: tests/AsyncRgbLedTestDriver.cpp [hunk 7]
```diff
@@ -0,0 +1,335 @@
+#include "TestInstance.h"
+
+#include "MockChannelData.h"
+#include "MockResults.h"
+#include "MockSettings.h"
+
+#include <cassert>
+#include <iostream>
```
Comment (Marcus10110) [inline tests/AsyncRgbLedTestDriver.cpp]:
```
#include <cstring> //for strcmp
```

File: CMakeLists.txt [hunk 8]
```diff
@@ -44,7 +44,17 @@ target_include_directories(AsyncRgbLedAnalyzer PRIVATE ${ANALYZER_SDK_ROOT}/incl
...
-set(CMAKE_EXPORT_COMPILE_COMMANDS ON)
+#------------------------------------------------------------------------
+# Testing
+
+add_subdirectory(AnalyzerTestHarness)
+
+add_executable(AsyncRgbLedTest tests/AsyncRgbLedTestDriver.cpp ${SOURCES})
+target_link_libraries(AsyncRgbLedTest AnalyzerTestHarness)
 
+add_test(AsyncRgbLedTest ${EXECUTABLE_OUTPUT_PATH}/${name})
```
Comment (Marcus10110) [inline CMakeLists.txt]: This doesn't seem to work on my machine, and I don't know where ${name} comes from. I changed this to:
```
add_test(AsyncRgbLedTest AsyncRgbLedTest)
```
and I added `enable_testing()` above. then `make test` ran successfully on Linux.

**Review by Marcus10110 — CHANGES_REQUESTED** [review]
included are comments to restore building on gcc and MSVC.

There is one other problem on MSVC.

The following classes are marked with `LOGICAPI`:
```
AnalyzerSettingInterfaceText
AnalyzerSettingInterfaceBool
AnalyzerSettingInterfaceInteger
```
However, none of their members are implemented. MSVC gives linker errors for the destructor and the `GetType` methods on each of these. I'm not 100% sure why, but removing `LOGICAPI` from their definitions fixes the errors and allows the project to link and for the tests to run and pass.
`LOGICAPI` expands to `__declspec(dllexport)` on MSVC.
We could just implement these methods to get it to link, or find another solution.

File: AnalyzerTestHarness/SettingsStubs.cpp (hunk 1 above)
Comment (zakalawe) [reply AnalyzerTestHarness/SettingsStubs.cpp]: Done

File: AnalyzerTestHarness/MockSimulatedChannelDescriptor.cpp (hunk 2 above)
Comment (zakalawe) [reply AnalyzerTestHarness/MockSimulatedChannelDescriptor.cpp]: Done with a stub for now, will be expanded when we test simulated data.

File: AnalyzerTestHarness/MockSimulatedChannelDescriptor.cpp (hunk 3 above)
Comment (zakalawe) [reply AnalyzerTestHarness/MockSimulatedChannelDescriptor.cpp]: Ditto.

File: AnalyzerTestHarness/HelperStubs.cpp (hunk 4 above)
Comment (zakalawe) [reply AnalyzerTestHarness/HelperStubs.cpp]: Copied the (two line) impl from Logic's AnalyzerHelpers.cpp

File: AnalyzerTestHarness/HelperStubs.cpp (hunk 5 above)
Comment (zakalawe) [reply AnalyzerTestHarness/HelperStubs.cpp]: Stubbed for now, until we do real testing of settings archival

**Review by Marcus10110 — APPROVED** [review]

**Comment by Marcus10110** [comment]
just caught up with the latest changes, looks great!

File: AnalyzerTestHarness/TestInstance.cpp (hunk 6 above)
Comment (Marcus10110) [reply AnalyzerTestHarness/TestInstance.cpp]: this cstring is still missing, and breaking the build with gcc.

**Comment by zakalawe** [comment]
@Marcus10110 do you want me to land these changes on master? Or hold off a bit longer? Will reply to other comments via email later

**Comment by Marcus10110** [comment]
@zakalawe I changed the base of this PR to `topics/analysis` so that the "Files Changed" view would only reflect changes since the last PR.
Ultimately let's get everything into master, which we can fast-forward once testing is finished.

**Review by Marcus10110 — CHANGES_REQUESTED** [review]
Builds and tests pass on MacOS and Linux.
On MSVC, there are a few issues - hannah found these too and included fixes for them here: 
https://github.com/TheOneRing/AsyncRgbLedAnalyzer/commit/b829efc6d94ce51ab45d2ef41d100700b5599424
However I suggest using either the C++11 new random features to replace `drand48()`.
Also now I don't think adding NOMINMAX to the defines in cmake was necessary to fix the uses of `std::max` and `std::min`, I think that might have just needed `#include <algorithm>`
Once this builds and tests pass out of the box on all 3 platforms, we can get it merged!

**Comment by zakalawe** [comment]
@Marcus10110 Hopefully this passes on Linux and Windows now? I didn't address the rpath issue however

**Comment by Marcus10110** [comment]
@zakalawe `#include <algorithm>` is still missing from `AsyncRgbLedTestDriver.cpp` which breaks the build on MSVC. After adding that, it builds and passes on MSVC.

**Review by Marcus10110 — APPROVED** [review]
woot! this builds and passes on all 3 platforms, merging to master!

**Comment by zakalawe** [comment]
Wonderful!

**Pull request merged**